crates/replace/src/lib.rs

```diff
@@ -605,22 +610,46 @@ impl<T: Config> Module<T> {
         // only cancellable by new_vault
         ensure!(caller == new_vault_id, Error::<T>::UnauthorizedVault);
 
-        // Release the newVault's collateral associated with this ReplaceRequests
-        ext::collateral::release_collateral::<T>(new_vault_id.clone(), replace.collateral)?;
-
-        // Transfer the oldVault’s griefing collateral associated with this ReplaceRequests to the newVault by calling slashCollateral
-        ext::collateral::slash_collateral::<T>(
-            replace.old_vault.clone(),
-            new_vault_id.clone(),
-            replace.griefing_collateral,
+        // decrease old-vault's to-be-redeemed tokens, and
+        // decrease new-vault's to-be-issued tokens
+        ext::vault_registry::cancel_replace_tokens::<T>(
+            &replace.old_vault,
+            &new_vault_id,
+            replace.amount,
         )?;
 
-        // Call the decreaseToBeRedeemedTokens function in the VaultRegistry for the oldVault.
-        let tokens = replace.amount;
-        ext::vault_registry::decrease_to_be_redeemed_tokens::<T>(
-            replace.old_vault.clone(),
-            tokens,
-        )?;
+        // slash old-vault's griefing collateral
+        if !ext::vault_registry::is_vault_liquidated::<T>(&new_vault_id)? {
+            // new-vault is not liquidated - give it the griefing collateral
+            ext::vault_registry::slash_collateral::<T>(
+                CurrencySource::Griefing(replace.old_vault.clone()),
+                CurrencySource::Backing(new_vault_id.clone()),
+                replace.griefing_collateral,
+            )?;
+        } else {
+            // new-vault is liquidated - slash to its free balance
+            ext::vault_registry::slash_collateral::<T>(
+                CurrencySource::Griefing(replace.old_vault.clone()),
+                CurrencySource::FreeBalance(new_vault_id.clone()),
+                replace.griefing_collateral,
+            )?;
+        }
+
+        // if the new_vault locked additional collateral especially for this replace,
+        // release it if it does not cause him to be undercollateralized
+        if !ext::vault_registry::is_vault_liquidated::<T>(&new_vault_id)? {
+            let new_collateral = ext::vault_registry::get_backing_collateral::<T>(&new_vault_id)?
+                .checked_sub(&replace.collateral)
+                .ok_or(Error::<T>::ArithmeticUnderflow)?;
+
+            let required_collateral =
+                ext::vault_registry::get_required_collateral_for_vault::<T>(&new_vault_id)?;
+
+            if new_collateral >= required_collateral {
+                // Release the newVault's collateral associated with this ReplaceRequests
+                ext::vault_registry::_withdraw_collateral::<T>(&new_vault_id, replace.collateral)?;
```

Comment:
This might be weird behavior if the new_vault specified additional collateral for the replace.

Reply:
To me it makes sense - the additional collateral in accept_replace is specifically to back the PolkaBTC it will receive from the old_vault. If the replace is cancelled, it should therefore be withdrawn. If it wants to increase its collateral for general purpose use, it should manually call `lock_additional_collateral` imo